File: src/client/camera.c
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <math.h>

#include "camera.h"
#include "window.h"

void camera_update(Camera *camera, float dt) {
  float yaw, pitch;
  window_getYawPitch(&yaw, &pitch);
  bool *keys = window_getKeys();

  // Update camera rotation
  camera->direction->x = -yaw;
  camera->direction->y = -pitch;

  // Update camera position
  {
    int forward = keys[WKEY_W] - keys[WKEY_S];
    int left = keys[WKEY_A] - keys[WKEY_D];
    int up = keys[WKEY_SPACE] - keys[WKEY_SHIFTL];

    camera->position->z += (forward * cos(camera->direction->x) -
                            left * sin(camera->direction->x)) *
                           *camera->speed * dt;
    camera->position->x += (left * cos(camera->direction->x) +
                            forward * sin(camera->direction->x)) *
                           *camera->speed * dt;
    camera->position->y += up * *camera->speed * dt;
  }

  // Wireframe
  if (window_wasPressed(WKEY_P)) camera->polygon_mode = !camera->polygon_mode;
}

File: src/client/camera.zig
const c = @cImport({
    @cInclude("RGFW.h");
});

const std = @import("std");
const math = std.math;

const Vec3 = extern struct { x: f32, y: f32, z: f32 };

pub const Camera = extern struct {
    position: *Vec3,
    direction: *Vec3,
    speed: *f32,
};

pub export fn camera_update(camera: *Camera, keys: [*]c_char, dx: f32, dy: f32, dt: f32) void {
    // Update camera rotation
    {
        const up: f32 = @as(f32, @floatFromInt(keys[c.RGFW_Up] - keys[c.RGFW_Down]));
        const left: f32 = @as(f32, @floatFromInt(keys[c.RGFW_Left] - keys[c.RGFW_Right]));

        camera.direction.y += up * dt;
        camera.direction.x += left * dt;
        camera.direction.y += -dy * dt * 0.15;
        camera.direction.x += -dx * dt * 0.15;

        camera.direction.y = @max(@min(camera.direction.y, math.pi / 2.0), -math.pi / 2.0);
    }

    // Update camera position
    {
        const forward: f32 = @as(f32, @floatFromInt(keys[c.RGFW_w] - keys[c.RGFW_s]));
        const left: f32 = @as(f32, @floatFromInt(keys[c.RGFW_a] - keys[c.RGFW_d]));
        const up: f32 = @as(f32, @floatFromInt(keys[c.RGFW_Space] - keys[c.RGFW_ShiftL]));

        camera.position.z += (forward * math.cos(camera.direction.x) - left * math.sin(camera.direction.x)) * camera.speed.* * dt;
        camera.position.x += (left * math.cos(camera.direction.x) + forward * math.sin(camera.direction.x)) * camera.speed.* * dt;
        camera.position.y += up * camera.speed.* * dt;
    }
}
